lib/cancel-stuck-txs.js

```diff
@@ -0,0 +1,144 @@
+      }
+    }
+  )
+  if (!res.ok) {
+    throw new Error(`Filfox request failed with ${res.status}: ${(await res.text()).trimEnd()}`)
+  }
+  /** @type {{
+     messages: {
+      cid: string;
+      from: string;
+      to: string;
+      nonce: number;
+      value: string;
+      gasLimit: number;
+      gasFeeCap: string;
+      gasPremium: string;
+      method: string;
+      methodNumber: number;
+      evmMethod: string;
+      createTimestamp: number;
+    }[];
+  }}
+  */
+  const { messages } = /** @type {any} */(await res.json())
+  return messages
+}
+
+/**
+ * @returns {Promise<{
+  "cid": string;
+  "height": number;
+  "timestamp": number;
+  "gasLimit": number;
+  "gasFeeCap": string;
+  "gasPremium": string;
+  "method": string;
+  "methodNumber": number;
+  "receipt": {
+      "exitCode": number;
+      "return": string;
+      "gasUsed": number;
+    },
+    "size": number;
+    "error": string;
+    "baseFee": string;
+    "fee": {
+      "baseFeeBurn": string;
+      "overEstimationBurn": string;
+      "minerPenalty": string;
+      "minerTip": string;
+      "refund": string;
+    },
+  }>}
+*/
+async function getRecentSendMessage () {
+  let res = await fetch('https://filfox.info/api/v1/message/list?method=Send')
+  if (!res.ok) {
+    throw new Error(`Filfox request failed with ${res.status}: ${(await res.text()).trimEnd()}`)
+  }
+  const body = /** @type {any} */(await res.json())
+  assert(body.messages.length > 0, '/message/list returned an empty list')
+  const sendMsg = body.messages.find(m => m.method === 'Send')
+  assert(!!sendMsg, 'No Send message found in the recent committed messages')
+  const cid = sendMsg.cid
+
+  res = await fetch(`https://filfox.info/api/v1/message/${cid}`)
+  if (!res.ok) {
+    throw new Error(`Filfox request failed with ${res.status}: ${(await res.text()).trimEnd()}`)
+  }
+
+  return /** @type {any} */(await res.json())
+}
+
+export const cancelStuckTxs = async ({ walletDelegatedAddress, signer }) => {
+  console.log('Checking for stuck transactions...')
+
+  const messages = await getMessagesInMempool(walletDelegatedAddress)
+  const txsToCancel = messages.filter(m => {
+    return new Date(m.createTimestamp * 1000) <
+      new Date(Date.now() - ROUND_LENGTH_MS)
+  })
+  if (txsToCancel.length === 0) {
+    console.log('No transactions to cancel')
+    return false
+  }
+
+  console.log('Going to revoke:')
+  for (const tx of txsToCancel) {
+    console.log('-', tx.cid, new Date(tx.createTimestamp * 1000).toISOString())
+  }
+
+  const recentSendMessage = await getRecentSendMessage()
+  console.log('Calculating gas fees from the recent Send message %s (created at %s)',
+    recentSendMessage.cid,
+    new Date(recentSendMessage.timestamp * 1000).toISOString()
+  )
+
+  const gasUsed = recentSendMessage.receipt.gasUsed
+  const gasFeeCap = Number(recentSendMessage.gasFeeCap)
+
+  await Promise.all(txsToCancel.map(async tx => {
```

Comment:
Try to cancel as many as possible in parallel, therefore await transaction confirmation. If we went serial (with a recently measured cancel duration of 13 minutes), we would likely never catch up with the queue. And if we didn't await confirmation, we might try to cancel the same transaction multiple times (fix: keep an in memory array of pending cancellations). Wdyt @bajtos?

Reply:
I've been running this script on my machine for a bit and the selected approach seems to work fine :)

Reply:
I am happy with the approach you implemented 👍🏻